File: packages/modules/secure-store/android/src/main/java/run/stingjs/modules/securestore/SecureStoreModule.kt
package run.stingjs.modules.securestore

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import java.nio.charset.StandardCharsets
import java.security.KeyStore
import java.util.concurrent.Executors
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import run.stingjs.runtime.StingNativeModule
import run.stingjs.runtime.StingNativeModuleCompletion
import run.stingjs.runtime.StingNativeModuleError
import run.stingjs.runtime.StingNativeModuleResult

class SecureStoreModule(context: Context) : StingNativeModule {
    override val name = "SecureStore"
    override val version = "0.1.0"

    private val preferences = context.applicationContext.getSharedPreferences(
        "sting-secure-store",
        Context.MODE_PRIVATE,
    )
    private val executor = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "StingSecureStore").apply { isDaemon = true }
    }

    override fun callSync(method: String, arguments: List<Any?>): Any? {
        throw StingNativeModuleError(
            code = "E_SYNC_UNSUPPORTED",
            message = "SecureStore methods are asynchronous",
        )
    }

    override fun callAsync(
        method: String,
        arguments: List<Any?>,
        completion: StingNativeModuleCompletion,
    ) {
        executor.execute {
            try {
                completion(StingNativeModuleResult.Success(perform(method, arguments)))
            } catch (error: StingNativeModuleError) {
                completion(StingNativeModuleResult.Failure(error))
            } catch (error: Throwable) {
                completion(
                    StingNativeModuleResult.Failure(
                        StingNativeModuleError(
                            code = "E_SECURE_STORE",
                            message = error.message ?: "SecureStore operation failed",
                            details = mapOf("method" to method),
                        ),
                    ),
                )
            }
        }
    }

    private fun perform(method: String, arguments: List<Any?>): Any? = when (method) {
        "getItem" -> {
            val key = storageKey(arguments)
            preferences.getString(key, null)?.let(::decrypt)
        }
        "setItem" -> {
            val key = storageKey(arguments, namespaceIndex = 2)
            val value = arguments.getOrNull(1) as? String
                ?: throw invalidArgument("SecureStore.setItem requires a string value")
            if (value.isEmpty()) throw invalidArgument("SecureStore.setItem requires a non-empty string value")
            if (!preferences.edit().putString(key, encrypt(value)).commit()) {
                throw StingNativeModuleError(
                    code = "E_SECURE_STORE",
                    message = "SecureStore failed to persist value",
                )
            }
            null
        }
        "deleteItem" -> {
            val key = storageKey(arguments)
            if (!preferences.edit().remove(key).commit()) {
                throw StingNativeModuleError(
                    code = "E_SECURE_STORE",
                    message = "SecureStore failed to delete value",
                )
            }
            null
        }
        "hasItem" -> preferences.contains(storageKey(arguments))
        else -> throw StingNativeModuleError(
            code = "E_METHOD_NOT_FOUND",
            message = "SecureStore does not implement asynchronous method $method",
        )
    }

    private fun storageKey(arguments: List<Any?>, namespaceIndex: Int = 1): String {
        val key = requireString(arguments, 0, "key")
        val namespace = requireString(arguments, namespaceIndex, "namespace")
        return "${encodePart(namespace)}.${encodePart(key)}"
    }

    private fun requireString(arguments: List<Any?>, index: Int, label: String): String {
        val value = arguments.getOrNull(index) as? String
            ?: throw invalidArgument("SecureStore $label must be a non-empty string")
        if (value.isEmpty()) throw invalidArgument("SecureStore $label must be a non-empty string")
        return value
    }

    private fun encodePart(value: String): String = Base64.encodeToString(
        value.toByteArray(StandardCharsets.UTF_8),
        Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING,
    )

    private fun encrypt(value: String): String {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, secretKey())
        val ciphertext = cipher.doFinal(value.toByteArray(StandardCharsets.UTF_8))
        val iv = Base64.encodeToString(cipher.iv, Base64.NO_WRAP)
        val payload = Base64.encodeToString(ciphertext, Base64.NO_WRAP)
        return "$iv:$payload"
    }

    private fun decrypt(stored: String): String {
        val separator = stored.indexOf(':')
        if (separator <= 0 || separator == stored.lastIndex) {
            throw StingNativeModuleError(
                code = "E_SECURE_STORE",
                message = "SecureStore contains an invalid encrypted payload",
            )
        }
        val iv = Base64.decode(stored.substring(0, separator), Base64.NO_WRAP)
        val ciphertext = Base64.decode(stored.substring(separator + 1), Base64.NO_WRAP)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, secretKey(), GCMParameterSpec(128, iv))
        return String(cipher.doFinal(ciphertext), StandardCharsets.UTF_8)
    }

    private fun secretKey(): SecretKey {
        val keyStore = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        (keyStore.getKey(KEY_ALIAS, null) as? SecretKey)?.let { return it }

        val generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
        generator.init(
            KeyGenParameterSpec.Builder(
                KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT,
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setRandomizedEncryptionRequired(true)
                .build(),
        )
        return generator.generateKey()
    }

    private fun invalidArgument(message: String) = StingNativeModuleError(
        code = "E_INVALID_ARGUMENT",
        message = message,
    )

    companion object {
        private const val KEY_ALIAS = "run.stingjs.secure-store.v1"
    }
}

File: packages/modules/secure-store/ios/SecureStoreModule.swift
import Foundation
import Security
import StingRuntime

public final class SecureStoreModule: StingNativeModule {
    public let name = "SecureStore"
    public let version = "0.1.0"

    private let queue = DispatchQueue(label: "run.stingjs.modules.secure-store", qos: .utility)

    public init() {}

    public func callSync(method: String, arguments: [Any]) throws -> Any? {
        throw StingNativeModuleError(
            code: "E_SYNC_UNSUPPORTED",
            message: "SecureStore methods are asynchronous"
        )
    }

    public func callAsync(
        method: String,
        arguments: [Any],
        completion: @escaping StingNativeModuleCompletion
    ) {
        queue.async { [self] in
            do {
                completion(.success(try perform(method: method, arguments: arguments)))
            } catch let error as StingNativeModuleError {
                completion(.failure(error))
            } catch {
                completion(.failure(StingNativeModuleError(
                    code: "E_SECURE_STORE",
                    message: error.localizedDescription,
                    details: ["method": method]
                )))
            }
        }
    }

    private func perform(method: String, arguments: [Any]) throws -> Any? {
        switch method {
        case "getItem":
            let (key, namespace) = try keyAndNamespace(arguments)
            return try read(key: key, namespace: namespace)
        case "setItem":
            guard arguments.count >= 3, let value = arguments[1] as? String, !value.isEmpty else {
                throw invalidArgument("SecureStore.setItem requires a non-empty string value")
            }
            let key = try requireString(arguments, index: 0, label: "key")
            let namespace = try requireString(arguments, index: 2, label: "namespace")
            try write(value: value, key: key, namespace: namespace)
            return nil
        case "deleteItem":
            let (key, namespace) = try keyAndNamespace(arguments)
            try delete(key: key, namespace: namespace)
            return nil
        case "hasItem":
            let (key, namespace) = try keyAndNamespace(arguments)
            return try exists(key: key, namespace: namespace)
        default:
            throw StingNativeModuleError(
                code: "E_METHOD_NOT_FOUND",
                message: "SecureStore does not implement asynchronous method \(method)"
            )
        }
    }

    private func keyAndNamespace(_ arguments: [Any]) throws -> (String, String) {
        (
            try requireString(arguments, index: 0, label: "key"),
            try requireString(arguments, index: 1, label: "namespace")
        )
    }

    private func requireString(_ arguments: [Any], index: Int, label: String) throws -> String {
        guard index < arguments.count, let value = arguments[index] as? String, !value.isEmpty else {
            throw invalidArgument("SecureStore \(label) must be a non-empty string")
        }
        return value
    }

    private func service(_ namespace: String) -> String {
        "run.stingjs.secure-store.\(namespace)"
    }

    private func baseQuery(key: String, namespace: String) -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service(namespace),
            kSecAttrAccount as String: key,
        ]
    }

    private func read(key: String, namespace: String) throws -> String? {
        var query = baseQuery(key: key, namespace: namespace)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne

        var result: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        if status == errSecItemNotFound { return nil }
        try requireSuccess(status, operation: "read")

        guard let data = result as? Data, let value = String(data: data, encoding: .utf8) else {
            throw StingNativeModuleError(
                code: "E_SECURE_STORE",
                message: "SecureStore value is not valid UTF-8"
            )
        }
        return value
    }

    private func write(value: String, key: String, namespace: String) throws {
        let data = Data(value.utf8)
        let query = baseQuery(key: key, namespace: namespace)
        let updateStatus = SecItemUpdate(
            query as CFDictionary,
            [kSecValueData as String: data] as CFDictionary
        )

        if updateStatus == errSecItemNotFound {
            var add = query
            add[kSecValueData as String] = data
            add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
            try requireSuccess(SecItemAdd(add as CFDictionary, nil), operation: "write")
            return
        }
        try requireSuccess(updateStatus, operation: "write")
    }

    private func delete(key: String, namespace: String) throws {
        let status = SecItemDelete(baseQuery(key: key, namespace: namespace) as CFDictionary)
        if status == errSecItemNotFound { return }
        try requireSuccess(status, operation: "delete")
    }

    private func exists(key: String, namespace: String) throws -> Bool {
        var query = baseQuery(key: key, namespace: namespace)
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        query[kSecReturnAttributes as String] = true
        let status = SecItemCopyMatching(query as CFDictionary, nil)
        if status == errSecItemNotFound { return false }
        try requireSuccess(status, operation: "inspect")
        return true
    }

    private func requireSuccess(_ status: OSStatus, operation: String) throws {
        guard status == errSecSuccess else {
            let message = SecCopyErrorMessageString(status, nil) as String? ?? "OSStatus \(status)"
            throw StingNativeModuleError(
                code: "E_SECURE_STORE",
                message: "SecureStore \(operation) failed: \(message)",
                details: ["status": Int(status)]
            )
        }
    }

    private func invalidArgument(_ message: String) -> StingNativeModuleError {
        StingNativeModuleError(code: "E_INVALID_ARGUMENT", message: message)
    }
}
